docs/pallas/tpu/matmul.md

```diff
@@ -0,0 +1,643 @@
+
+> Note that the only way of allocating scratch spaces right now is via `pltpu.PrefetchScalarGridSpec`. Don't worry about exactly what it does for now -- all you need to know for now is that it allows you to allocate scratch spaces in VMEM.
+
+```{code-cell} ipython3
+:id: tyMcZtA6dWDP
+
+def matmul_kernel(x_ref, y_ref, z_ref, acc_ref, *, nsteps):
+  @pl.when(pl.program_id(2) == 0)
+  def _():
+    acc_ref[...] = jnp.zeros_like(acc_ref)
+
+  acc_ref[...] += jnp.dot(
+      x_ref[...], y_ref[...], preferred_element_type=jnp.float32
+  )
+
+  @pl.when(pl.program_id(2) == nsteps - 1)
+  def _():
+    z_ref[...] = acc_ref[...].astype(z_ref.dtype)
+
+
+@functools.partial(jax.jit, static_argnames=['bm', 'bk', 'bn'])
+def matmul(
+    x: jax.Array,
+    y: jax.Array,
+    *,
+    bm: int = 128,
+    bk: int = 128,
+    bn: int = 128,
+):
+  m, k = x.shape
+  _, n = y.shape
+  return pl.pallas_call(
+      functools.partial(matmul_kernel, nsteps=k // bk),
+      grid_spec=pltpu.PrefetchScalarGridSpec(
+        num_scalar_prefetch=0,
+        in_specs=[
+            pl.BlockSpec(lambda i, j, k: (i, k), (bm, bk)),
+            pl.BlockSpec(lambda i, j, k: (k, j), (bk, bn)),
+        ],
+        out_specs=pl.BlockSpec(lambda i, j, k: (i, j), (bm, bn)),
+        scratch_shapes=[pltpu.VMEM((bm, bn), jnp.float32)],
+        grid=(m // bm, n // bn, k // bk),
+      ),
+      out_shape=jax.ShapeDtypeStruct((m, n), x.dtype),
+  )(x, y)
+```
+
+```{code-cell} ipython3
+:id: G3uHKEabVXep
+
+m, k, n = 4096, 4096, 4096
+k1, k2 = random.split(random.key(0), 2)
+x = random.normal(k1, (m, k), dtype=jnp.bfloat16)
+y = random.normal(k2, (k, n), dtype=jnp.bfloat16)
+np.testing.assert_array_equal(x @ y, matmul(x, y))
+```
+
++++ {"id": "fBL1NwXzVlWa"}
+
+## Performance of pipelined kernels
+
+Our above analysis about FLOPS vs memory usage applies at a coarse scale i.e. when we are looking at the the size of a the total matrix multiplication. However, remember that in practice, we are pipelining the execution of a blocked matrix multiplication, meaning we have a loop in which we are doing matrix multiplication with smaller blocks.
+
+This means that we actually care about the FLOPS vs memory bandwidth usage of each individual instance of the kernel, not the global FLOPS vs memory bandwidth usage. Therefore, the block sizes `bm`, `bk`, `bn` are extremely important for matrix multiplication performance. Even if we have the largest matrices in the world, if we pick very small `bm`, `bk`, and `bn`, we will be memory bound because each time we invoke the kernel we will have too few FLOPs to hide the memory transfers happening in the background.
```

Comment:
block sizes are important for all workloads, not just this one :) I'd remove "for matrix multiplication performance", because you've already said matmul too many times haha

Reply:
There's another important point that we haven't made here: the larger the blocks, the smaller the total number of bytes transferred! This is why you want to make the blocks as large as possible, up to low-level memory capacity limits. The real formula for computing the total number of flops transferred by a tiled matmul is this:
```
((m * k) * (n // bn) + (k * n) * (m // bm) + n * m) * bytewidth
```

Reply:
I guess this doesn't matter as much if each iteration we are compute bound (since we will always be doing the minimum amount of FLOPs). We don't mind re-reading inputs in that case.